MobilityData/mobility-feed-api
# feat: automate gtfs & gbfs data preprocessing for analytics

**Summary:**

This PR introduces the implementation of two Google Cloud Functions, `process_analytics_gtfs` and `process_analytics_gbfs`, which automate the monthly retrieval, processing, and storage of analytics data for GTFS and GBFS datasets. These functions are scheduled to run once a month via GCP Scheduler and store the resulting analytics in the specified Google Cloud Storage buckets.

The changes include the creation of base and specific analytics processors, responsible for handling the data retrieval, processing metrics, and saving the results in JSON format. 

**Expected behavior:**

- Upon execution, the `process_analytics_gtfs` function retrieves the latest GTFS dataset per feed, processes it to extract metrics related to validation notices, features, and geographical locations, and stores the results as JSON files in the designated GCS bucket.
- Similarly, the `process_analytics_gbfs` function retrieves the latest GBFS snapshot per feed, processes it to extract metrics related to validation notices, versions, and geographical locations, and stores the results as JSON files in the appropriate GCS bucket.

Both functions will also update the `analytics_files.json` index in the storage buckets, ensuring that all analytics files are appropriately listed.

**Testing tips:**

1. Verify that the analytics files are correctly generated and stored in the development storage buckets:
     - GTFS: [GTFS Analytics in Dev](https://console.cloud.google.com/storage/browser/mobilitydata-gtfs-analytics-dev;tab=objects?forceOnBucketsSortingFiltering=true&project=mobility-feeds-dev&prefix=&forceOnObjectsSortingFiltering=false)
     - GBFS: [GBFS Analytics in Dev](https://console.cloud.google.com/storage/browser/mobilitydata-gbfs-analytics-dev;tab=objects?forceOnBucketsSortingFiltering=true&project=mobility-feeds-dev&prefix=&forceOnObjectsSortingFiltering=false)

2. Execute the functions manually to ensure they are working as expected and check the execution logs for any errors:
     - GTFS:
       ```bash
       curl -m 550 -X POST https://northamerica-northeast1-mobility-feeds-dev.cloudfunctions.net/process-analytics-gtfs \
       -H "Authorization: bearer $(gcloud auth print-identity-token)" \
       -H "Content-Type: application/json" \
       -d '{
           "compute_date": "20240901"
       }'
       ```
     - GBFS:
       ```bash
       curl -m 550 -X POST https://northamerica-northeast1-mobility-feeds-dev.cloudfunctions.net/process-analytics-gbfs \
       -H "Authorization: bearer $(gcloud auth print-identity-token)" \
       -H "Content-Type: application/json" \
       -d '{
           "compute_date": "20240901"
       }'
       ```

3. After execution, review the logs to ensure the functions processed the data correctly without any issues and that the metrics files were properly updated in the GCS buckets.


Please make sure these boxes are checked before submitting your pull request - thanks!
